agent/mapcall/src/main/java/randoop/instrument/MapCallsAgent.java

```diff
@@ -1,322 +1,195 @@
-
-      String bcel_classname = "org.apache.bcel.Constants";
-      String pag_marker_classname = "org.apache.bcel.PAGMarker";
-
-      List<URL> bcel_urls = get_resource_list(bcel_classname);
-      List<URL> pag_urls = get_resource_list(pag_marker_classname);
-
-      if (pag_urls.isEmpty()) {
-        System.err.printf(
-            "%nBCEL must be in the classpath.  " + "Normally it is found in daikon.jar .%n");
-        System.exit(1);
-      }
-      if (bcel_urls.size() < pag_urls.size()) {
-        System.err.printf("%nCorrupted BCEL library, bcel %s, pag %s%n", bcel_urls, pag_urls);
+    // If user provided package exclusion file, load user package exclusions
+    if (dont_transform != null) {
+      try {
+        loadExclusions(
+            new FileReader(dont_transform), dont_transform.getName(), excludedPackagePrefixes);
+      } catch (IOException e) {
+        System.err.format(
+            "Error reading package exclusion file %s:%n %s%n", dont_transform, e.getMessage());
         System.exit(1);
       }
-
-      // No need to do anything if only our versions of bcel are present
-      if (bcel_urls.size() == pag_urls.size()) return;
-
-      int bcel_index = 0;
-      int pag_index = 0;
-      while (bcel_index < bcel_urls.size()) {
-        URL bcel = bcel_urls.get(bcel_index);
-        URL pag = pag_urls.get(pag_index);
-        if (!pag.getProtocol().equals("jar")) {
-          System.err.printf("%nPAG BCEL must be in jar file. " + " Found at %s%n", pag);
-          System.exit(1);
-        }
-        if (same_location(bcel, pag)) {
-          if (bcel_index == pag_index) {
-            URL first_bcel = bcel;
-            while ((pag != null) && same_location(bcel, pag)) {
-              bcel = bcel_urls.get(++bcel_index);
-              pag_index++;
-              pag = (pag_index < pag_urls.size()) ? pag_urls.get(pag_index) : null;
-            }
-            System.err.printf(
-                "%nPAG BCEL (%s) appears before target BCEL "
-                    + "(%s).%nPlease reorder classpath to put randoop.jar at "
-                    + "the end.%n",
-                first_bcel, bcel);
-            System.exit(1);
-          } else {
-            bcel_jar = new JarFile(extract_jar_path(pag));
-            debug.log("PAG BCEL found in jar %s%n", bcel_jar.getName());
-            break;
-          }
-        } else { // non pag bcel found
-          debug.log("Found non-pag BCEL at %s%n", bcel);
-          bcel_index++;
-        }
-      }
     }
 
-    /**
-     * Returns whether or not the two URL represent the same location for org.apache.bcel. Two
-     * locations match if they refer to the same jar file or the same directory in the filesystem.
-     */
-    private boolean same_location(URL url1, URL url2) {
-      if (!url1.getProtocol().equals(url2.getProtocol())) {
-        return false;
-      }
+    ConcurrentHashMap<MethodSignature, MethodSignature> replacementMap = new ConcurrentHashMap<>();
```

Comment:
When Randoop runs with threads, the agent can run in more than one and it is necessary to allow concurrent access to the replacement map.

The `MethodReplacements` class is providing the signature strings of the replaced methods to Randoop, and does not use a map. The strings are used to avoid Randoop being dependent on the agent.